### PR TITLE
Multiples and full cover

### PR DESCRIPTION
**Pull request by nulty**
# Purpose

Add the ability to calculate the maximum returns for bets such as a Trixie (Multiple) and a Patent (Full Cover), with unlimited selections.

# Approach

To reduce code and leverage the power of es2015, as much of this implementation has been made using functional code. as a result there are fewer lines of code than might otherwise be required. 

# Concerns
The API for the single and accumulator functions return an object whereas the API for the multiple function returns a single number. The differences between these two return values are potentially confusing.

The reason for the return value being an object with the `single` and `accumulator` is so the place stake can be applied to a bet independently.

- In a Single eachway bet, the Win stake and the place stake are the same, i.e. the original stake.
But when an eachway Double is made, we initially pass the original stake to the Single function first, then using the returns from that bet, we must use the Win result of the first bet as the Win stake for the second part of the Double and we must use the Place result of the first bet as the Place stake for the second part of the Double. These two values need to be kept discreet.

In order for this to be possible, the only way I could think to achieve this with recursion was to return the bet returns individually in this object.

Suggestions on a better approach are most welcome.

# Tests
```
$ babel-node test/test.js
TAP version 13
# single win bet is correct
ok 1 should be equal
# single eachway is correct
ok 2 should be equal
# win double is correct
ok 3 should be equal
# eachway double is correct
ok 4 should be equal
# multiple with two selections
ok 5 should be equal
# ew multiple with two selections
ok 6 should be equal
# full cover multiple with three selections
ok 7 should be equal
# full cover ew multiple with three selections
ok 8 should be equal

1..8
# tests 8
# pass  8

# ok
```

### CODE REVIEW

File: lib/combinations.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+function combinations(arr, size) {
+  var len = arr.length;
```
Comment (inicklas) [inline lib/combinations.js]: since this isn't being reassigned, can we declare it with `const`?

File: lib/combinations.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  if (size > len) return [];
+  if (!size) return [[]];
+  if (size == len) return [arr];
+
+  return arr.reduce(function (acc, val, i) {
+    var res = combinations(arr.slice(i + 1), size - 1)
+    .map(function (comb) { return [val].concat(comb); });
+
+    return acc.concat(res);
```
Comment (revdan) [inline lib/combinations.js]: A couple of things in here. I don't think `res` needs to be a var - maybe let or const. 

I think the formatting needs some indentation, or a new variable on the second line of the reduce block. That could also use the es6 arrow syntax nicely as well `res.map((comb) => [val].concat(comb));`

File: lib/combinations.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+function combinations(arr, size) {
+  var len = arr.length;
+
+  if (size > len) return [];
+  if (!size) return [[]];
+  if (size == len) return [arr];
+
+  return arr.reduce(function (acc, val, i) {
+    var res = combinations(arr.slice(i + 1), size - 1)
```
Comment (inicklas) [inline lib/combinations.js]: can this be changed to `let`?

File: lib/main.js [hunk 4]
```diff
@@ -1,38 +1,57 @@
...
-    function bet(stake, odds) {
-      return (stake * (odds - 1)) + stake;
-    }
+function bet(stake, odds) {
+  return (stake * (odds - 1)) + stake;
+}
 
-    function single(stake, odds, terms, placeStake = 0) {
-      const newPlaceStake = terms && placeStake === 0 ? stake : placeStake;
+function single(stake, odds, terms, ew, placeStake = 0) {
```
Comment (revdan) [inline lib/main.js]: I think it makes sense to have named params for the arguments, so the user can be more sure they're providing the correct values and not worrying about ordering

**Review by revdan — COMMENTED** [review]
Looks good!

File: lib/main.js (hunk 4 above)
Comment (inicklas) [inline lib/main.js]: what type is `ew` and can it be anything?. 

Typescript would be nice 🤔  (Just saying)

Also I was wondering if explicitly using json key value pairs will help anyone know what is expected for this `function`?

**Review by inicklas — COMMENTED** [review]
i really really like the tests. Very self explanatory. 👍 I just added a few pointers we can think about.

File: lib/combinations.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+function combinations(arr, size) {
+  var len = arr.length;
+
+  if (size > len) return [];
+  if (!size) return [[]];
+  if (size == len) return [arr];
```
Comment (hopeshigh) [inline lib/combinations.js]: can these be converted to a standard if block?
```
if (size > len) {
  return [];
}
```

File: lib/main.js [hunk 6]
```diff
@@ -1,38 +1,57 @@
...
 
-      const placeReturn = terms ? bet(newPlaceStake, ((odds - 1) * terms) + 1) : 0;
+  const placeReturn = ew ? bet(newPlaceStake, ((odds - 1) * terms) + 1) : 0;
 
-      const winReturn = bet(stake, odds);
+  const winReturn = bet(stake, odds);
 
-      return { win: winReturn, place: placeReturn, total: (winReturn + placeReturn) };
-    }
+  return { win: winReturn, place: placeReturn, total: (winReturn + placeReturn) };
```
Comment (hopeshigh) [inline lib/main.js]: this return maybe more readable as:
```
return {
  win: winReturn, 
  place: placeReturn, 
  total: (winReturn + placeReturn)
}
```

File: lib/main.js [hunk 7]
```diff
@@ -1,38 +1,57 @@
...
-    function multiple(selections, stake, idx = 0, place) {
-      const sel = selections[idx];
+function accumulator(selections, stake, ew, idx = 0, place = 0) {
+  const sel = selections[idx];
 
-      const res = single(stake, sel.odds, sel.terms, place);
+  const res = single(stake, sel.odds, sel.terms, ew, place);
 
-      if (idx + 1 === selections.length) return res.total;
+  if (idx + 1 === selections.length) return res.total;
```
Comment (hopeshigh) [inline lib/main.js]: same with changing the one liner

**Review by hopeshigh — COMMENTED** [review]
Some comments.

File: dist/bet-return.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+function multiple(selections, stake, ew = false, fullCover = false) {
+  const count = selections.length;
+  const start = fullCover ? 1 : 2;
+  let allCombos = [];
+
+  for (let i = start; i <= count; i += 1) {
+    allCombos = allCombos.concat(combinations(selections, i));
+  }
```
Comment (inicklas) [inline dist/bet-return.js:57]: I don't know if this reads better. See if you like it?

```javascript

selections.forEach((_selection, index) => {
  allCombos = allCombos.concat(combinations(selections, index + 1));
}
```

**Review by inicklas — APPROVED** [review]
🚀 good job sir!

File: dist/bet-return.js (hunk 8 above)
Comment (nulty) [reply dist/bet-return.js:57]: That won't work afaik. The `i` will start at either 1 or 2 depending on whether we are calculating multiple or full cover..... as in if we want to calculate the singles as well as the multiples

File: lib/combinations.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+function combinations(collection, size) {
+  const len = collection.length;
+
+  if (size > len) return [];
+  if (!size) return [[]];
+  if (size === len) return [collection];
```
Comment (hopeshigh) [inline lib/combinations.js:6]: I would convert these from one liners to:
if (foo) { THING HERE }

File: lib/main.js [hunk 10]
```diff
@@ -1,38 +1,54 @@
...
+  };
+}
 
-      const placeReturn = terms ? bet(newPlaceStake, ((odds - 1) * terms) + 1) : 0;
+function accumulator(selections, stake, ew, idx = 0, place = 0) {
+  const sel = selections[idx];
 
-      const winReturn = bet(stake, odds);
+  const res = single({
+    stake, odds: sel.odds, terms: sel.terms, ew, placeStake: place
```
Comment (hopeshigh) [inline lib/main.js]: I would have this split over multiple lines as it's an object

File: lib/main.js [hunk 11]
```diff
@@ -1,38 +1,54 @@
...
+  let allCombos = [];
 
-      if (idx + 1 === selections.length) return res.total;
+  for (let i = start; i <= count; i += 1) {
+    allCombos = allCombos.concat(combinations(selections, i));
+  }
 
-      return multiple(selections, res.win, idx + 1, res.place || 0);
-    };
+  return allCombos.reduce((acc, selection) => acc + accumulator(selection, stake, ew), 0);
```
Comment (hopeshigh) [inline lib/main.js]: to make this bit a bit nicer to read, maybe do a similar declaration of allCombos = THIS then you can just have a nice line at the end with return allCombos.

**Review by hopeshigh — COMMENTED** [review]
Took me a while to realise I had reviewed the dist/index.js.

Suggested some thoughts

File: lib/combinations.js (hunk 9 above)
Comment (nulty) [reply lib/combinations.js:6]: This is technically undefined in airbnb styleguide though I'm happy to have it up for discussion.

http://airbnb.io/javascript/#blocks

Comment (hopeshigh) [reply lib/combinations.js:6]: Apologies, the styling ruined my point, I meant like this part:
```
// good
if (test) {
  return false;
}
```

**Pull request merged**